networkx/algorithms/bipartite/projection.py

```diff
@@ -185,33 +185,72 @@ def weighted_projected_graph(B, nodes, ratio=False):
         Networks". In Carrington, P. and Scott, J. (eds) The Sage Handbook
         of Social Network Analysis. Sage Publications.
     """
-    if B.is_directed():
-        pred = B.pred
-        G = nx.DiGraph()
+
+    if type(nodes) != list:
+        top_nodes = []
+        for node in B.nodes():
+            if type(node) != type(nodes):
+                top_nodes.append(node)
+            elif node not in nodes:
+                top_nodes.append(node)
     else:
-        pred = B.adj
-        G = nx.Graph()
-    G.graph.update(B.graph)
-    G.add_nodes_from((n, B.nodes[n]) for n in nodes)
-    n_top = len(B) - len(nodes)
+        top_nodes = [node for node in B.nodes() if node not in nodes]
 
-    if n_top < 1:
+    if len(top_nodes) + len(nodes) != B.number_of_nodes():
         raise NetworkXAlgorithmError(
             f"the size of the nodes to project onto ({len(nodes)}) is >= the graph size ({len(B)}).\n"
             "They are either not a valid bipartite partition or contain duplicates"
         )
 
-    for u in nodes:
-        unbrs = set(B[u])
-        nbrs2 = {n for nbr in unbrs for n in B[nbr]} - {u}
-        for v in nbrs2:
-            vnbrs = set(pred[v])
-            common = unbrs & vnbrs
-            if not ratio:
-                weight = len(common)
+    if not B.is_directed():
+        # Get the biadjacency matrix.
+        adj = nx.bipartite.biadjacency_matrix(
+            B, row_order=nodes, column_order=top_nodes
+        )
+        G = nx.Graph()
+        G.add_nodes_from(nodes)
+
+        # Get the co-citation matrix.
+        projected_adj = adj @ adj.T
+        projected_adj.setdiag(0)
+
+        # Add the edges from the result.
+        denom = 1 / len(top_nodes)
+        for row, col in zip(*projected_adj.nonzero()):
+            val = projected_adj[row, col]
+            if ratio:
+                G.add_edge(nodes[row], nodes[col], weight=val * denom)
             else:
-                weight = len(common) / n_top
-            G.add_edge(u, v, weight=weight)
+                G.add_edge(nodes[row], nodes[col], weight=val)
```

Comment:
If I'm understanding this correctly, I think you could significantly improve performance (and readability) futher by using array math, e.g. something like

```python
projected_adj = projected_adj.to_array()
if ratio:
    projected_adj = projected_adj / len(top_nodes)
return nx.from_numpy_array(projected_adj)
```